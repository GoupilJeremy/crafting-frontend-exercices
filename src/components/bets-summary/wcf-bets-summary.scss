@import '../../shared/styles/index.scss';

.bets-summary {
    @extend %main-wrapper;
    @extend %separator;

    h3 {
        @extend %medium-title;
    }

    &__info {
        @extend %content-wrapper;

        p {
            @extend %medium-text;
        }
    }
}

.summary__value {
    font-weight: $text-bold;
}