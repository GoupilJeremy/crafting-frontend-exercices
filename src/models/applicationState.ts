import { GameInfo, OddsInfo, Bet } from '.';

export type ApplicationState = {
    gameInfos: GameInfo[];
    oddsInfo: OddsInfo[];
    selectedBets: Bet[];
}
