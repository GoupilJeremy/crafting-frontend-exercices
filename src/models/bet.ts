import { GameInfo } from "./game";
import { OddsInfo } from "./odds";

export type BetInfo = Omit<GameInfo & OddsInfo, 'id'> & {
    [key: string]: number;
};

export type BetChoice = '1' | 'draw' | '2';
export type Bet = {
    gameId: string;
    selectedChoice: BetChoice;
    selectedOdd: number;
}