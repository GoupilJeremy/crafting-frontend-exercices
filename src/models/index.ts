export * from './games'
export * from './api'
export * from './bets'
export * from './storybook'