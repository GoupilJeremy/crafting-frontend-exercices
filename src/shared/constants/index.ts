export * from './events'
export * from './betChoice'