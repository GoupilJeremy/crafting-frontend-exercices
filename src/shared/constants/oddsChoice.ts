export const CHOICE_LEFT = '1';
export const CHOICE_DRAW = 'draw';
export const CHOICE_RIGHT = '2';