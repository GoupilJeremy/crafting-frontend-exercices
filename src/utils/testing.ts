export function mockPropsTo(element: Element) {
    return function (key: string, value: unknown) {
        element.setAttribute(key, JSON.stringify(value))
    }
}

export function dispatchMockedEventWith(element: Element) {
    return function (key: string, value: number | string) {
        const inputElement = element.shadowRoot?.querySelector('input')!;
        inputElement.value = value.toString();

        const fakeEvent = new Event(key)
        inputElement.dispatchEvent(fakeEvent);
    }
}

export function findElementWith(element: Element, key: string): Element | null | undefined {
    return element.shadowRoot?.querySelector(key);
}

export function isVisible(element: Element | null | undefined): boolean {
    if (!element) return false;

    return !element.hasAttribute('hidden');
}