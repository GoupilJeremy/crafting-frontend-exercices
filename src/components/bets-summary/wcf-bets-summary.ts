import css from './wcf-bets-summary.scss';
import { getPotentialGain } from "../../business/bets/getPotentialGain";
import { getSumOfOdds } from "../../business/odds/getSumOfOdds";
import { Bet } from "../../models/bet";

const template = document.createElement('template');
template.innerHTML = `
<style>${css}</style>
<div class="bets-summary">
    <h3 class="bets-summary__title">Récapitulatif de vos paris</h3>
    <div class="bets-summary__info">
        <p class="bets-summary__info--bets-number"></p>
        <p class="bets-summary__info--sum-odds"></p>
        <p class="bets-summary__info--potential-gain"></p>
    </div>
</div>
`;

export class BetsSummary extends HTMLElement {
    private bets: Bet[] = [];
    private startingBet: number = 0;
    private isUserPrenium: boolean = false;

    constructor() {
        super();

        this.attachShadow({ mode: 'open' })
            .appendChild(template.content.cloneNode(true));
    }

    connectedCallback() {
        this.displaySummaryInfo()
    }

    static get observedAttributes() {
        return ['bets', 'startingbet', 'isuserprenium']
    }

    attributeChangedCallback(name: string, _oldValue: string, newValue: string) {
        switch (name) {
            case 'bets':
                this.bets = JSON.parse(newValue);
                break;
            case 'startingbet':
                this.startingBet = Number(newValue);
                break;
            case 'isuserprenium':
                this.isUserPrenium = Boolean(newValue);
                break;
            default:
                break;
        }

        if (this.startingBet && this.bets.length) {
            this.displaySummaryInfo();
        }
    }

    displaySummaryInfo() {
        this.shadowRoot!.querySelector('.bets-summary__info--bets-number')!.textContent = `Nombre de paris joués: ${this.bets.length}`
        this.shadowRoot!.querySelector('.bets-summary__info--sum-odds')!.textContent = `Côte(s) cummulée(s): ${getSumOfOdds(this.bets)}`
        this.shadowRoot!.querySelector('.bets-summary__info--potential-gain')!.textContent = `Potentiel gain: ${getPotentialGain(this.startingBet, this.bets, this.isUserPrenium)}`
    }
}