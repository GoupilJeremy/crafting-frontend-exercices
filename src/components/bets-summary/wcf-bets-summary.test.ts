import { Bet } from "../../models/bet";
import { CHOICE_1, CHOICE_DRAW } from "../../shared/constants/oddsChoice";
import { mockPropsTo } from "../../utils/testing";
import { BetsSummary } from "./wcf-bets-summary";

const DUMMY_BETS: Bet[] = [
    {
        gameId: 'id1',
        selectedChoice: CHOICE_1,
        selectedOdd: 1.52,
    },
    {
        gameId: 'id2',
        selectedChoice: CHOICE_DRAW,
        selectedOdd: 1.12,
    }
];

let betsSummary: BetsSummary;
let addProps: Function;

describe('Bets-Summary Component', () => {
    beforeEach(() => {
        // When
        betsSummary = new BetsSummary();

        addProps = mockPropsTo(betsSummary)
    })

    it('should render correctly info when user is NOT premium', () => {
        // Given
        addProps('bets', DUMMY_BETS)
        addProps('startingbet', 100)

        // When
        betsSummary.connectedCallback();

        // Then
        expect(betsSummary?.shadowRoot?.innerHTML).toMatchSnapshot();
    });

    it('should render correctly info when user is premium', () => {
        // Given
        addProps('bets', DUMMY_BETS)
        addProps('startingbet', 100)
        addProps('isuserpremium', true)

        // When
        betsSummary.connectedCallback();

        // Then
        expect(betsSummary?.shadowRoot?.innerHTML).toMatchSnapshot();
    });
});
