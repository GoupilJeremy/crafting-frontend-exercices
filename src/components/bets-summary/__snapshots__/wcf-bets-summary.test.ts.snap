// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bets-Summary Component should render correctly info when user is NOT prenium 1`] = `
"
<style>[object Object]</style>
<div class=\\"bets-summary\\">
    <h3 class=\\"bets-summary__title\\">Récapitulatif de vos paris</h3>
    <div class=\\"bets-summary__info\\">
        <p class=\\"bets-summary__info--bets-number\\">Nombre de paris joués: 2</p>
        <p class=\\"bets-summary__info--sum-odds\\">Côte(s) cummulée(s): 1.7</p>
        <p class=\\"bets-summary__info--potential-gain\\">Potentiel gain: 170</p>
    </div>
</div>
"
`;

exports[`Bets-Summary Component should render correctly info when user is prenium 1`] = `
"
<style>[object Object]</style>
<div class=\\"bets-summary\\">
    <h3 class=\\"bets-summary__title\\">Récapitulatif de vos paris</h3>
    <div class=\\"bets-summary__info\\">
        <p class=\\"bets-summary__info--bets-number\\">Nombre de paris joués: 2</p>
        <p class=\\"bets-summary__info--sum-odds\\">Côte(s) cummulée(s): 1.7</p>
        <p class=\\"bets-summary__info--potential-gain\\">Potentiel gain: 187</p>
    </div>
</div>
"
`;
